File: rust/introsort.rs
// Introsort implementation in Rust (generic, in-place)
pub fn introsort<T: Ord>(arr: &mut [T]) {
    if arr.len() <= 1 { return; }
    let depth_limit = 2 * ((arr.len() as f64).log2().floor() as usize);
    introsort_impl(arr, depth_limit);
}

fn introsort_impl<T: Ord>(arr: &mut [T], mut depth_limit: usize) {
    const INSERTION_THRESHOLD: usize = 16;

    let mut low = 0usize;
    let mut high = arr.len();

    // Use a manual stack to avoid deep recursion on tail calls
    let mut stack: Vec<(usize, usize, usize)> = Vec::new();
    stack.push((low, high, depth_limit));

    while let Some((lo, hi, depth)) = stack.pop() {
        let len = hi.saturating_sub(lo);
        if len <= 1 { continue; }

        if len <= INSERTION_THRESHOLD {
            insertion_sort(&mut arr[lo..hi]);
            continue;
        }

        if depth == 0 {
            heapsort(&mut arr[lo..hi]);
            continue;
        }

        // Partition with median-of-three pivot
        let pivot_index = partition_median3(&mut arr[lo..hi]);
        // pivot_index is relative to the subslice [lo..hi]
        // push larger side first so smaller side is processed next (keeps stack small)
        let left_len = pivot_index;
        let right_len = len - pivot_index - 1;

        if left_len < right_len {
            // push right, then left
            if pivot_index + 1 < len {
                stack.push((lo + pivot_index + 1, hi, depth - 1));
            }
            if left_len > 0 {
                stack.push((lo, lo + pivot_index, depth - 1));
            }
        } else {
            if left_len > 0 {
                stack.push((lo, lo + pivot_index, depth - 1));
            }
            if pivot_index + 1 < len {
                stack.push((lo + pivot_index + 1, hi, depth - 1));
            }
        }
    }
}

fn insertion_sort<T: Ord>(slice: &mut [T]) {
    for i in 1..slice.len() {
        let mut j = i;
        while j > 0 && slice[j] < slice[j - 1] {
            slice.swap(j, j - 1);
            j -= 1;
        }
    }
}

fn partition_median3<T: Ord>(slice: &mut [T]) -> usize {
    let len = slice.len();
    let mid = len / 2;
    // choose indices 0, mid, len-1 -> median-of-three
    let (a, b, c) = (0usize, mid, len - 1);
    // reorder such that slice[a] <= slice[b] <= slice[c]
    if slice[a] > slice[b] { slice.swap(a, b); }
    if slice[b] > slice[c] { slice.swap(b, c); }
    if slice[a] > slice[b] { slice.swap(a, b); }

    // move pivot to end-1 (so we'll partition excluding last element)
    slice.swap(b, len - 1);
    let pivot_idx = len - 1;
    let mut i = 0usize;
    for j in 0..(len - 1) {
        if slice[j] <= slice[pivot_idx] {
            slice.swap(i, j);
            i += 1;
        }
    }
    slice.swap(i, pivot_idx);
    i
}

fn heapsort<T: Ord>(slice: &mut [T]) {
    let len = slice.len();
    if len <= 1 { return; }

    // build max heap
    for start in (0..=(len/2)).rev() {
        sift_down(slice, start, len);
    }

    // sortdown
    for end in (1..len).rev() {
        slice.swap(0, end);
        sift_down(slice, 0, end);
    }

    fn sift_down<T: Ord>(slice: &mut [T], start: usize, end: usize) {
        let mut root = start;
        loop {
            let left = 2 * root + 1;
            if left >= end { break; }
            let mut swap = root;
            if slice[swap] < slice[left] {
                swap = left;
            }
            let right = left + 1;
            if right < end && slice[swap] < slice[right] {
                swap = right;
            }
            if swap == root { break; }
            slice.swap(root, swap);
            root = swap;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_introsort_simple() {
        let mut v = vec![3, 1, 4, 1, 5, 9, 2, 6, 5];
        introsort(&mut v);
        assert_eq!(v, {
            let mut t = vec![1,1,2,3,4,5,5,6,9];
            t
        });
    }

    #[test]
    fn test_introsort_sorted() {
        let mut v: Vec<i32> = (0..100).collect();
        introsort(&mut v);
        assert_eq!(v, (0..100).collect::<Vec<i32>>());
    }

    #[test]
    fn test_introsort_reverse() {
        let mut v: Vec<i32> = (0..100).rev().collect();
        introsort(&mut v);
        assert_eq!(v, (0..100).collect::<Vec<i32>>());
    }
}

File: go/intro_sort.go
package main

import (
	"fmt"
	"math"
)

// Introsort entry point (generic for any ordered type)
func IntroSort[T constraints.Ordered](arr []T) {
	if len(arr) <= 1 {
		return
	}
	depthLimit := int(2 * math.Floor(math.Log2(float64(len(arr)))))
	introsortImpl(arr, depthLimit)
}

func introsortImpl[T constraints.Ordered](arr []T, depthLimit int) {
	const INSERTION_THRESHOLD = 16

	n := len(arr)
	if n <= 1 {
		return
	}

	if n <= INSERTION_THRESHOLD {
		insertionSort(arr)
		return
	}

	if depthLimit == 0 {
		heapSort(arr)
		return
	}

	// Partition using median-of-three pivot
	pivot := partitionMedian3(arr)
	introsortImpl(arr[:pivot], depthLimit-1)
	introsortImpl(arr[pivot+1:], depthLimit-1)
}

func insertionSort[T constraints.Ordered](arr []T) {
	for i := 1; i < len(arr); i++ {
		for j := i; j > 0 && arr[j] < arr[j-1]; j-- {
			arr[j], arr[j-1] = arr[j-1], arr[j]
		}
	}
}

func partitionMedian3[T constraints.Ordered](arr []T) int {
	n := len(arr)
	mid := n / 2

	// median-of-three: arr[0], arr[mid], arr[n-1]
	if arr[0] > arr[mid] {
		arr[0], arr[mid] = arr[mid], arr[0]
	}
	if arr[mid] > arr[n-1] {
		arr[mid], arr[n-1] = arr[n-1], arr[mid]
	}
	if arr[0] > arr[mid] {
		arr[0], arr[mid] = arr[mid], arr[0]
	}

	// pivot = arr[mid]
	pivot := arr[mid]
	arr[mid], arr[n-1] = arr[n-1], arr[mid] // move pivot to end
	i := 0
	for j := 0; j < n-1; j++ {
		if arr[j] < pivot {
			arr[i], arr[j] = arr[j], arr[i]
			i++
		}
	}
	arr[i], arr[n-1] = arr[n-1], arr[i]
	return i
}

func heapSort[T constraints.Ordered](arr []T) {
	n := len(arr)
	// Build max-heap
	for i := n/2 - 1; i >= 0; i-- {
		siftDown(arr, i, n)
	}
	// Extract elements
	for i := n - 1; i > 0; i-- {
		arr[0], arr[i] = arr[i], arr[0]
		siftDown(arr, 0, i)
	}
}

func siftDown[T constraints.Ordered](arr []T, start, end int) {
	root := start
	for {
		child := 2*root + 1
		if child >= end {
			break
		}
		swap := root
		if arr[swap] < arr[child] {
			swap = child
		}
		if child+1 < end && arr[swap] < arr[child+1] {
			swap = child + 1
		}
		if swap == root {
			return
		}
		arr[root], arr[swap] = arr[swap], arr[root]
		root = swap
	}
}
